src/tactic/unfold_cases.lean

```diff
@@ -0,0 +1,169 @@
+  we do not get these stronger lemmas from the compiler for free but must in fact
+  prove them using `cases` or some other local reasoning.
+
+  In other words, the following does not constitute a proof that lean accepts.
+  ```lean
+  example : ∀ x, myand ff x = ff :=
+  begin
+    intros, refl
+  end
+  ```
+
+  However, you can use `unfold_cases { refl }` to prove `∀ x, myand ff x = ff` and
+  `∀ x, myand x ff = ff`. For definitions with many cases, the savings can be very
+  significant.
+
+  The term that gets generated for the above definition looks like this:
+  ```lean
+  λ (a a_1 : bool),
+  a.cases_on
+    (a_1.cases_on (id_rhs bool ff) (id_rhs bool ff))
+    (a_1.cases_on (id_rhs bool ff) (id_rhs bool tt))
+  ```
+
+  When the tactic tries to prove the goal `∀ x, myand ff x = ff`, it starts by `intros`,
+  followed by unfolding the definition:
+  ```lean
+  ⊢ ff.cases_on
+    (x.cases_on (id_rhs bool ff) (id_rhs bool ff))
+    (x.cases_on (id_rhs bool ff) (id_rhs bool tt)) = ff
+  ```
+
+  At this point, it can make progress using `dsimp`. But then it gets stuck:
+  ```lean
+  ⊢ bool.rec (id_rhs bool ff) (id_rhs bool ff) x = ff
+  ```
+
+  Next, it can introduce a case split on `x`. At this point, it has to prove two
+  goals:
+  ```lean
+  ⊢ bool.rec (id_rhs bool ff) (id_rhs bool ff) ff = ff
+  ⊢ bool.rec (id_rhs bool ff) (id_rhs bool ff) tt = ff
+  ```
+
+  Now, however, both goals can be discharged using `refl`.
+-/
+namespace tactic
+open expr
+namespace unfold_cases
+/--
+  Given an equation `f x = y`, this tactic tries to infer an expression that can be
+  used to do distinction by cases on to make progress.
+
+  Pre-condition: assumes that the outer-most application cannot be beta-reduced
+  (e.g. `whnf` or `dsimp`).
+-/
+meta def find_splitting_expr : expr → tactic expr
+| `(@ite %%cond %%dec_inst _ _ _ = _) := pure `(@decidable.em %%cond %%dec_inst)
+| `(%%l@(app x y) = _) := pure y
+| e := fail!"expected an expression of the form: f x = y. Got:\n{e}"
+
+/--
+  Tries to finish the current goal using the `inner` tactic. If the tactic
+  fails, it tries to find an expression on which to do a distinction by
+  cases and calls itself recursively.
+
+  The order of operations is significant. Because the unfolding can potentially
+  be infinite, it is important to apply the `inner` tactic at every step.
+
+  Notice, that if the `inner` tactic succeeds, the recursive unfolding is stopped.
+-/
+meta def unfold_cases_core (inner : interactive.itactic) : tactic unit :=
+inner <|>
+(do split_ifs [], all_goals unfold_cases_core, skip) <|>
+do
+  tgt ← target,
+  e ← find_splitting_expr tgt,
+  focus1 $ do
+    cases e,
+    all_goals $ (dsimp_target >> unfold_cases_core) <|> skip,
+    skip
+
+/--
+  Given a target of the form `⊢ f x₁ ... xₙ = y`, unfolds `f` using a delta reduction.
+-/
+meta def unfold_tgt : expr → tactic unit
+| e@`(%%l@(app _ _) = %%r) :=
```

Comment:
```suggestion
| `(%%l@(app _ _) = %%r) :=
```
This binding is never used since it gets shadowed by other `e`s below; I think it just makes things slightly harder to read.